### PR TITLE
Densenet canonizations

### PR DESCRIPTION
**Pull request by gumityolcu**
Hello,

Here are a summary of the contributions:

1. The epsilon values of the batch_norm layers used to be left as they are, when they need to be set to zero for perfect canonization. I fixed this. Note that the epsilon parameter can not be added to the batch_norm_params field of the canonizer. This is because it is a literal, not a torch Variable. So if you add the epsilon parameter there, the code will try to reach batch_norm.eps.data, which does not exits, when trying to restore it. Therefore, i use a new class variable "batch_norm_eps" to remember it during canonization.
2. CompositeCanonizer now returns the list of handles **reversed**. This is because, if we have a two canonizers attaching to a module, then we need to detach them in the reverse order that they are applied, in order to restore the original values. I opted to reverse the list in the class because detaching the given handles in returned order seemed more user friendly. And I couldn't think of a use case where this would cause problems.
3. MergeBatchNormtoRight canonizer is added. This merges a batch normalization layer to a linear layer that comes **after** it. If the linear layer is a convolutional layer with padding, this is not straightforward. A full feature map needs to be added to the output of the layer instead of a simple bias. This is done by adding forward hooks.
4. ThreshReLUMergeBatchNorm is added. This canonizer detects BN->ReLU->Linear and changes the activation function to a function that depends on the batch norm variables to get the BatchNorm after the activation. Then the batchnorm is merged to the linear layer that is next to it. This is as described in https://github.com/AlexBinder/LRP_Pytorch_Resnets_Densenet/blob/master/canonization_doc.pdf

Further more BN->ReLU->AvgPool->Linear chains are found and canonized using the same method, because Batch normalization commutes with average pooling. 
6.Full proposed canonizers are added to torchvision.py. Another addition is DenseNetAdaptiveAvgPoolCanonizer which is needed before applying other canonizers to densenets. It makes the final ReLU and AvgPooling layers of torchvision densenet objects explicit. By default, these are applied in the forward method of the model, not as nn.module objects.

Thank you very much and I am looking forward to any kind of feedback!

### CODE REVIEW

File: src/zennit/canonizers.py [hunk 1]
```diff
@@ -181,13 +187,256 @@ def apply(self, root_module):
...
+    def prehook(module, x):
+        module.canonization_params["original_x"] = x[0].clone()
+
+    @staticmethod
+    def fwdhook(module, x, y):
+        x = module.canonization_params["original_x"]
+        index = (None, slice(None), *((None,) * (module.canonization_params['weights'].ndim + 1)))
+        y = module.canonization_params['weights'][index] * x + module.canonization_params['biases'][index]
+        baseline_vals = -1. * (module.canonization_params['biases'] / module.canonization_params['weights'])[index]
+        return torch.where(y > 0, x, baseline_vals)
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:372]: missing docstrings and signatures are non-descriptive

File: src/zennit/canonizers.py [hunk 2]
```diff
@@ -181,13 +187,256 @@ def apply(self, root_module):
...
+            bias_kernel = bias_kernel[:, :, indices1, :]
+            bias_kernel = bias_kernel[:, :, :, indices2]
+        ynew = y + bias_kernel
+        return ynew
+
+    def __init__(self):
+        super().__init__()
+        self.handles = []
+
+    def apply(self, root_module):
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:252]: missing docstring

File: src/zennit/canonizers.py [hunk 3]
```diff
@@ -181,13 +187,256 @@ def apply(self, root_module):
...
+        instances = []
+        oldest_leaf = None
+        old_leaf = None
+        mid_leaf = None
+        for leaf in collect_leaves(root_module):
+            if (
+                    isinstance(old_leaf, self.batch_norm_type)
+                    and isinstance(mid_leaf, ReLU)
+                    and isinstance(leaf, self.linear_type)
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:387]: this looks like one indentation level too deep

File: src/zennit/canonizers.py [hunk 4]
```diff
@@ -181,13 +187,256 @@ def apply(self, root_module):
...
+            ):
+                instance = self.copy()
+                instance.register((leaf,), old_leaf, mid_leaf)
+                instances.append(instance)
+            elif (
+                    isinstance(oldest_leaf, self.batch_norm_type)
+                    and isinstance(old_leaf, ReLU)
+                    and isinstance(mid_leaf, AdaptiveAvgPool2d)
+                    and isinstance(leaf, self.linear_type)
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:396]: this looks like one indentation level too deep

File: src/zennit/canonizers.py [hunk 5]
```diff
@@ -181,13 +187,256 @@ def apply(self, root_module):
...
+        index = (None, slice(None), *((None,) * (module.canonization_params['weights'].ndim + 1)))
+        y = module.canonization_params['weights'][index] * x + module.canonization_params['biases'][index]
+        baseline_vals = -1. * (module.canonization_params['biases'] / module.canonization_params['weights'])[index]
+        return torch.where(y > 0, x, baseline_vals)
+
+    def __init__(self):
+        super().__init__()
+        self.relu = None
+
+    def apply(self, root_module):
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:378]: missing docstring

File: src/zennit/canonizers.py [hunk 6]
```diff
@@ -181,13 +187,256 @@ def apply(self, root_module):
...
+        denominator = (batch_norm.running_var + batch_norm.eps) ** .5
+        scale = (batch_norm.weight / denominator)  # Weight of the batch norm layer when seen as a linear layer
+        shift = batch_norm.bias - batch_norm.running_mean * scale  # bias of the batch norm layer when seen as a linear layer
+        self.relu.canonization_params = {}
+        self.relu.canonization_params['weights'] = scale
+        self.relu.canonization_params['biases'] = shift
+
+        super().register(linears, batch_norm)
+        self.handles.append(self.relu.register_forward_pre_hook(ThreshReLUMergeBatchNorm.prehook))
+        self.handles.append(self.relu.register_forward_hook(ThreshReLUMergeBatchNorm.fwdhook))
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:431]: maybe we can do this with a wrapped .forward similar to how I suggested for the `MergeBatchNormtoRight`?

File: src/zennit/canonizers.py [hunk 7]
```diff
@@ -181,13 +187,256 @@ def apply(self, root_module):
...
+            The activation unit between the Batch Normalization and Linear modules.
+        '''
+        self.relu = relu
+
+        denominator = (batch_norm.running_var + batch_norm.eps) ** .5
+        scale = (batch_norm.weight / denominator)  # Weight of the batch norm layer when seen as a linear layer
+        shift = batch_norm.bias - batch_norm.running_mean * scale  # bias of the batch norm layer when seen as a linear layer
+        self.relu.canonization_params = {}
+        self.relu.canonization_params['weights'] = scale
+        self.relu.canonization_params['biases'] = shift
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:427]: let's store these inside the canonizer

File: src/zennit/canonizers.py [hunk 8]
```diff
@@ -181,13 +187,256 @@ def apply(self, root_module):
...
+        super().register(linears, batch_norm)
+        self.handles.append(self.relu.register_forward_pre_hook(ThreshReLUMergeBatchNorm.prehook))
+        self.handles.append(self.relu.register_forward_hook(ThreshReLUMergeBatchNorm.fwdhook))
+
+    def remove(self):
+        '''Undo the merge by reverting the parameters of both the linear and the batch norm modules to the state before
+        the merge.
+        '''
+        super().remove()
+        delattr(self.relu, "canonization_params")
```
Comment (chr5tphr) [inline src/zennit/canonizers.py:438]: If we store the params inside the canonizer, we do not need to delete them, thus we can simply remove this to use the parent class' remove

File: src/zennit/torchvision.py [hunk 9]
```diff
@@ -147,3 +147,68 @@ def __init__(self):
...
+        module.features.add_module('final_relu', ReLU(inplace=True))
+        module.features.add_module('adaptive_avg_pool2d', AdaptiveAvgPool2d((1, 1)))
+        super(DenseNetAdaptiveAvgPoolCanonizer, self).register(module, attributes)
+
+    def remove(self):
+        '''Remove the overloaded attributes. Note that functions are descriptors, and therefore not direct attributes
+        of instance, which is why deleting instance attributes with the same name reverts them to the original
+        function.
+        '''
+        self.module.features = Sequential(*list(self.module.features.children())[:-2])
```
Comment (chr5tphr) [inline src/zennit/torchvision.py:188]: If I remember correctly, you can slice `Sequential`, i.e. `self.module.feature = self.module.features[:-2]`

File: src/zennit/torchvision.py [hunk 10]
```diff
@@ -147,3 +147,68 @@ def __init__(self):
...
+        '''Copy this Canonizer.
+
+        Returns
+        -------
+        obj:`Canonizer`
+            A copy of this Canonizer.
+        '''
+        return DenseNetAdaptiveAvgPoolCanonizer()
+
+    def register(self, module, attributes):
```
Comment (chr5tphr) [inline src/zennit/torchvision.py:178]: missing docstring

File: src/zennit/torchvision.py [hunk 11]
```diff
@@ -147,3 +147,68 @@ def __init__(self):
...
+    def remove(self):
+        '''Remove the overloaded attributes. Note that functions are descriptors, and therefore not direct attributes
+        of instance, which is why deleting instance attributes with the same name reverts them to the original
+        function.
+        '''
+        self.module.features = Sequential(*list(self.module.features.children())[:-2])
+        for key in self.attribute_keys:
+            delattr(self.module, key)
+
+    def forward(self, x):
```
Comment (chr5tphr) [inline src/zennit/torchvision.py:192]: missing docstring

File: src/zennit/torchvision.py [hunk 12]
```diff
@@ -147,3 +147,68 @@ def __init__(self):
...
+        out = self.features(x)
+        out = torch.flatten(out, 1)
+        out = self.classifier(out)
+        return out
+
+
+class DenseNetSeqThreshCanonizer(CompositeCanonizer):
```
Comment (chr5tphr) [inline src/zennit/torchvision.py:199]: missing docstring

**Review by chr5tphr — CHANGES_REQUESTED** [review]
Hey Galip,

sorry for the very long hold-up. Let's try to finalize this. Ultimately, we need to rebase this. Maybe you can first introduce the changes and then rebase.
